test/testdata/lsp/code_actions/extract_variable/invalid.rb

```diff
@@ -0,0 +1,100 @@
+# typed: true
+# selective-apply-code-action: refactor.extract
+# assert-no-code-action: refactor.extract
+def example(x)
+#           ^ apply-code-action: [A] Extract Variable
+  xs = []
+# ^^ apply-code-action: [A] Extract Variable
+  xs = []
+#      ^ apply-code-action: [A] Extract Variable
+  xs.each { |x| return x.foo if x.foo}
+#                                ^^^^ apply-code-action: [A] Extract Variable
+  T.must(123)
+#   ^^^^ apply-code-action: [A] Extract Variable
+  T.must(123)
+# ^^^^^^ apply-code-action: [A] Extract Variable
+  return
+# ^^^^^^ apply-code-action: [A] Extract Variable
+end
+
+class A
+  attr_accessor :a
+
+  x = 1 + 1
+#     ^^^ apply-code-action: [A] Extract Variable
+  x = 1 + 1
+#       ^ apply-code-action: [A] Extract Variable
+  x = 1 + 1
+# ^ apply-code-action: [A] Extract Variable
+
+  def multiple_return_break_next
+    1.times do
+      if 1 == 2
+        break 1, 2
+#       ^^^^^^^^^^ apply-code-action: [A] Extract Variable
+      else
+        next 1, 2
+#       ^^^^^^^^^^ apply-code-action: [A] Extract Variable
+      end
+    end
+    return 1, 2
+#   ^^^^^^^^^^^ apply-code-action: [A] Extract Variable
+  end
+end
+
+  A.new.a &&= 1
+# ^^^^^^^ apply-code-action: [A] Extract Variable
+
+  A.new.a ||= 1
+# ^^^^^^^ apply-code-action: [A] Extract Variable
+
+  A.new.a *= 1
+# ^^^^^^^ apply-code-action: [A] Extract Variable
+
+a = T.let(1, T.untyped)
+
+  a &&= 1
+# ^ apply-code-action: [A] Extract Variable
```

Comment:
Not sure how important it is to you, but you may want to write a test in ProtocolTest.cc to assert that if you ask for a code action with a zero-length range that no code action gets applied (or does the test harness have a way to specify a zero-length range natively? I didn't think it did)